cloudwatch_metrics_collector/test/metrics_collector_test.cpp

```diff
@@ -0,0 +1,238 @@
+#include <cloudwatch_metrics_common/metric_service_factory.hpp>
+#include <cloudwatch_metrics_collector/metrics_collector.hpp>
+
+#include "test_common.hpp"
+
+#include <gtest/gtest.h>
+
+using namespace Aws::CloudWatchMetrics::Utils;
+using namespace Aws::CloudWatchMetrics;
+using metrics_statistics_msgs::msg::StatisticDataType;
+
+class MetricServiceMock : public MetricService
+{
+public:
+  MetricServiceMock(const std::string & metrics_namespace) : MetricService(
+    std::make_shared<MetricPublisher>(metrics_namespace, Aws::Client::ClientConfiguration()),
+    std::make_shared<MetricBatcher>()) {}
+
+  bool batchData(const MetricObject & data_to_batch)
+  {
+    metric_objects_.push_back(data_to_batch);
+    return true;
+  }
+
+  const std::vector<MetricObject> & GetMetricObjects()
+  {
+    return metric_objects_;
+  }
+
+private:
+  std::vector<MetricObject> metric_objects_;
+};
+
+class MetricServiceFactoryMock : public MetricServiceFactory
+{
+public:
+  MetricServiceFactoryMock(std::shared_ptr<MetricService> metric_service)
+    : mock_metric_service_(std::move(metric_service)) {}
+
+  std::shared_ptr<MetricService> createMetricService(
+    const std::string & /*metrics_namespace*/,
+    const Aws::Client::ClientConfiguration & /*client_config*/,
+    const Aws::SDKOptions & /*sdk_options*/,
+    const CloudWatchOptions & /*cloudwatch_option*/)
+  {
+    return mock_metric_service_;
+  }
+
+private:
+  std::shared_ptr<MetricService> mock_metric_service_;
+};
+
+class MetricsCollectorFixture : public ::testing::Test
+{
+  void SetUp()
+  {
+    mock_metric_service_ = std::make_shared<MetricServiceMock>(kTestMetricsNamespace);
+    mock_metric_service_factory_ = std::make_shared<MetricServiceFactoryMock>(mock_metric_service_);
+
+    metrics_collector_.Initialize(
+      kTestMetricsNamespace,
+      default_dimensions_,
+      60,
+      nullptr,
+      client_config_,
+      sdk_options_,
+      cloudwatch_options_,
+      topics_,
+      mock_metric_service_factory_);
+  }
+
+protected:
+  static constexpr char kTestMetricsNamespace[] = "test_namespace";
+
+  Aws::Client::ClientConfiguration client_config_;
+  Aws::SDKOptions sdk_options_;
+  Aws::CloudWatchMetrics::CloudWatchOptions cloudwatch_options_;
+  std::vector<TopicInfo> topics_;
+
+  std::map<std::string, std::string> default_dimensions_;
+  std::shared_ptr<MetricServiceMock> mock_metric_service_;
+  std::shared_ptr<MetricServiceFactoryMock> mock_metric_service_factory_;
+  MetricsCollector metrics_collector_;
+};
+
+constexpr char MetricsCollectorFixture::kTestMetricsNamespace[];
+
+TEST_F(MetricsCollectorFixture, RecordRosMonitoringMessages)
+{
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dist(0.0, 100.0);
+
+  std::array<ros_monitoring_msgs::msg::MetricData, 2> metric_data = {{
+    BasicMonitoringData(),
+    BasicMonitoringData()
+  }};
+  for (auto & data : metric_data) {
+    data.value = dist(gen);
+  }
```

Comment:
Cool, but what value does this random data generator add? 
I don't see it helping improve the coverage , on the downside we have a clumsy loop with nested conditions within a unit test. I would just create 2 `BasicMonitoringData` and set values to it and check if you are able to extract those values out. 

I like simple unit tests that don't try to do too much. Can we simplify further?

Reply:
I think the value is in testing a greater range of inputs, including garbage-in / garbage-out. It may be clumsy, but it's definitely not flaky, and that's the important part.

Reply:
Okay, I have removed the random number generation.